OSDOCS-11444: Adds low latency to MicroShift (#80236)
openshift/openshift-docs: modules/microshift-low-latency-config-yaml.adoc

.Prerequisites

* You installed the OpenShift CLI (`oc`).
* You have root access to the cluster.
* You made a copy of the provided `config.yaml.default` file in the `/etc/microshift/` directory, renaming it `config.yaml`.
//anything else? do we need to create the KubeletConfiguration first? or does this config populate that for us?

.Procedure

. Add the kubelet configuration to the {microshift-short} `config.yaml` file
+
.Example passthrough `kubelet` configuration
[source,yaml]
----
# ...
kubelet: <1>
  cpuManagerPolicy: static
  cpuManagerPolicyOptions:
    full-pcpus-only: "true"
  cpuManagerReconcilePeriod: 5s
  memoryManagerPolicy: Static
  topologyManagerPolicy: single-numa-node
  reservedSystemCPUs: 0-1
  reservedMemory:
  - limits:
      memory: 1100Mi <2>
    numaNode: 0
  kubeReserved:
    memory: 500Mi
  systemReserved:
    memory: 500Mi
  evictionHard: <3>
    imagefs.available: "15%" <4>
    memory.available: "100Mi" <5>
    nodefs.available: "10%" <6>
    nodefs.inodesFree: "5%" <7>
  evictionPressureTransitionPeriod: 0s
# ...
----

> //How many footnotes do we want to include? Any others?

Reviewer: I think these could be useful:
- cpuManagerPolicy
- memoryManagerPolicy
- full-pcpus-only

reservedSystemCPUs would be kind of must have as this is very important to dividing management workload and apps

> <1> If you change the CPU or memory managers in the kubelet configuration, you must remove files that cache the previous configuration. Restart to remove them automatically, or manually remove the `- /var/lib/kubelet/cpu_manager_state` and `- /var/lib/kubelet/memory_manager_state` files.

Reviewer: Reboot? Restarting MicroShift is not enough because it's handled by another service.
```suggestion
<1> If you change the CPU or memory managers in the kubelet configuration, you must remove files that cache the previous configuration. Reboot to remove them automatically, or manually remove the `- /var/lib/kubelet/cpu_manager_state` and `- /var/lib/kubelet/memory_manager_state` files.
```

Author: It's a style thing not use "reboot", see https://www.ibm.com/docs/en/ibm-style?topic=word-usage#reboot
Is "restart the host" OK?

<3> The `evictionHard` parameters define under which conditions the kubelet evicts pods. When you change the
default value of only one parameter for the `evictionHard` stanza, the default values of other parameters are not inherited and are set to zero. Provide all the threshold values even when you want to change just one.
<4> The `imagefs` is a filesystem that container runtimes use to store container images and container writable layers. In this example, the `evictionHard.imagefs.available` parameter means that the pod is evicted when the available space of the image filesystem is less than 15%.
<5> In this example, the `evictionHard.memory.available` parameter means that the pod is evicted when the available memory of the node drops below 100MiB.
<6> In this example, the `evictionHard.nodefs.available` parameter means that the pod is evicted when the main filesystem of the node has less than 10% available space.
<7> In this example, the `evictionHard.nodefs.inodesFree` parameter means that the pod is evicted when more than 15% of the node's main filesystem's inodes are in use.

> //How do we verify that this passed to the kubelet correctly?

Reviewer: Provided settings will appear in /var/lib/microshift/resources/kubelet/config/config.yaml

// Module included in the following assemblies:
//
// * microshift_configuring/microshift-low-latency.adoc

:_mod-docs-content-type: PROCEDURE
[id="microshift-low-latency-config-yaml_{context}"]
= Configuration kubelet parameters and values in {microshift-short}

The first step in enabling low latency to a {microshift-short} cluster is to add configurations to the {microshift-short} `config.yaml` file.

.Prerequisites

* You installed the OpenShift CLI (`oc`).
* You have root access to the cluster.

> * You made a copy of the provided `config.yaml.default` file in the `/etc/microshift/` directory, renaming it `config.yaml`.

Reviewer 2: Keep past tense throughout the sentence?
```suggestion
* You made a copy of the provided `config.yaml.default` file in the `/etc/microshift/` directory and renamed it `config.yaml`.
```

.Procedure

> . Add the kubelet configuration to the {microshift-short} `config.yaml` file

Reviewer 3: ```suggestion
. Add the kubelet configuration to the {microshift-short} `config.yaml` file.
```

> . Add the kubelet configuration to the {microshift-short} `config.yaml` file:

Reviewer 2: In the core OCP docs, one-step procedures take an asterisk. 

```suggestion
* Add the kubelet configuration to the {microshift-short} `config.yaml` file:
```

+
.Example passthrough `kubelet` configuration
[source,yaml]
----

> # ...

Reviewer 2: The OCP core docs [require the metadata in YAML code blocks](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#required-fields). Does the same apply for MicroShift?

Author: We haven't done this because we have so many of these, but fixing the Config book is on our list. And at once point, another reviewer told me we didn't always have to do this in certain cases (I'll have to find that comment again to tell you what  the reason was). Meanwhile, I'll leave this one, but created an issue to clarify for this docset. https://issues.redhat.com/browse/OSDOCS-11859

kubelet: <1>
  cpuManagerPolicy: static # <2>
  cpuManagerPolicyOptions:
    full-pcpus-only: "true" # <3>
  cpuManagerReconcilePeriod: 5s

>   memoryManagerPolicy: static # <4>

Reviewer: This must start with upper case. The cpu manager's starts with lower case...
```suggestion
  memoryManagerPolicy: Static # <4>
```

  topologyManagerPolicy: single-numa-node
  reservedSystemCPUs: 0-1 # <5>
  reservedMemory:
  - limits:
      memory: 1100Mi # <6>
    numaNode: 0
  kubeReserved:
    memory: 500Mi
  systemReserved:
    memory: 500Mi
  evictionHard: # <7>
    imagefs.available: "15%" # <8>
    memory.available: "100Mi" # <9>
    nodefs.available: "10%" # <10>
    nodefs.inodesFree: "5%" # <11>
  evictionPressureTransitionPeriod: 0s
# ...
----

> <1> If you change the CPU or memory managers in the kubelet configuration, you must remove files that cache the previous configuration. Restart the host to remove them automatically, or manually remove the `- /var/lib/kubelet/cpu_manager_state` and `- /var/lib/kubelet/memory_manager_state` files.

Reviewer: ```suggestion
<1> If you change the CPU or memory managers in the kubelet configuration, you must remove files that cache the previous configuration. Restart the host to remove them automatically, or manually remove the `/var/lib/kubelet/cpu_manager_state` and `/var/lib/kubelet/memory_manager_state` files.
```

> <2> The name of the policy to use. Valid values are `none` and `static`. Requires the `CPUManager` feature gate to be enabled. Default value is `none`.

Reviewer 2: >Requires the `CPUManager` feature gate to be enabled.

Should this be a prereq? The wording might be tricky, as it seems required for only a couple of parameters. But, it might prevent the user from stopping cold here and needing to go back and do something else. (I'd get annoyed!).

Author: Hmmm. @pmtk Should this note be here, or is this referring to an OCP CR we don't use?

<3> A set of `key=value` for setting extra options that fine tune the behavior of the `CPUManager` policies.  Default: null. Requires both the `CPUManager` and `CPUManagerPolicyOptions` feature gates to be enabled.

> <4> The name of the policy used by Memory Manager. Case-sensitive. Default value is `none`. Requires the `MemoryManager` feature gate to be enabled.

Reviewer 2: ```suggestion
<4> The name of the policy used by Memory Manager. Case-sensitive. The default value is `none`. Requires the `MemoryManager` feature gate to be enabled.
```

Reviewer 2: > Requires the `MemoryManager` feature gate to be enabled.

Same concern as above. Should this be a prereq?

Author: WDYT @pmtk?

> <5> Required. This parameter is essential to dividing the management and application workloads. Use this parameter to define a static CPU set for the host-level system and Kubernetes daemons, plus interrupts and timers. Then the rest of the CPUs on the system can be used exclusively for workloads. Use your TuneD profile to move the system and Kubernetes daemons, interrupts, and timers to the explicit CPU set defined by this option.

Reviewer 4: Do we want to mention that reservedSystemCPUs should be the inverse of the isolated CPUs? As both combined need to account for all CPUs on the system.

Reviewer: I think it's mentioned in other place that they need to be complementary. However, this is not so clear if you're offlining CPUs.
@ShaunaDiaz I just checked the parent tuned profile (cpu-partitioning) and it specifically checks if `isolated_cores` have `offline` CPUs in it.
However, cpu-partitioning is activated before our profile, it doesn't seem to break anything (first a core is isolated, then offlined).
But I cannot vouch that it'll not break something somewhere...

> <5> Required. The `reservedSystemCPUs` value must be the inverse of the offlined CPUs because both combined must account for all of the CPUs on the system. This parameter is essential to dividing the management and application workloads. Use this parameter to define a static CPU set for the host-level system and Kubernetes daemons, plus interrupts and timers. Then the rest of the CPUs on the system can be used exclusively for workloads.

Reviewer 2: This seems to be the format favored in the docs: `Required:`
```suggestion
<5> Required: The `reservedSystemCPUs` value must be the inverse of the offlined CPUs, because both values combined must account for all of the CPUs on the system. This parameter is essential to dividing the management and application workloads. Use this parameter to define a static CPU set for the host-level system and Kubernetes daemons, plus interrupts and timers. Then the rest of the CPUs on the system can be used exclusively for workloads.
```

<6> The value in `reservedMemory[0].limits.memory`, `1100` Mi in this example, is equal to `kubeReserved.memory` + `systemReserved.memory` + `evictionHard.memory.available`.

> <7> The `evictionHard` parameters define under which conditions the kubelet evicts pods. When you change the
> default value of only one parameter for the `evictionHard` stanza, the default values of other parameters are not inherited and are set to zero. Provide all the threshold values even when you want to change just one.

Reviewer 2: ```suggestion
<7> The `evictionHard` parameters define under which conditions the kubelet evicts pods. When you change the default value of only one parameter for the `evictionHard` stanza, the default values of other parameters are not inherited and are set to zero. Provide all the threshold values even when you want to change just one.
```

> <8> The `imagefs` is a filesystem that container runtimes use to store container images and container writable layers. In this example, the `evictionHard.imagefs.available` parameter means that the pod is evicted when the available space of the image filesystem is less than 15%.

Reviewer 2: ```suggestion
<8> The `imagefs` is a filesystem that container runtimes used to store container images and container writable layers. In this example, the `evictionHard.imagefs.available` parameter means that the pod is evicted when the available space of the image filesystem is less than 15%.
```

> <9> In this example, the `evictionHard.memory.available` parameter means that the pod is evicted when the available memory of the node drops below 100MiB.
> <10> In this example, the `evictionHard.nodefs.available` parameter means that the pod is evicted when the main filesystem of the node has less than 10% available space.
> <11> In this example, the `evictionHard.nodefs.inodesFree` parameter means that the pod is evicted when more than 15% of the node's main filesystem's inodes are in use.

Reviewer: ```suggestion
<9> In this example, the `evictionHard.memory.available` parameter means that the Pods are evicted when the available memory of the node drops below 100MiB.
<10> In this example, the `evictionHard.nodefs.available` parameter means that the Pods are evicted when the main filesystem of the node has less than 10% available space.
<11> In this example, the `evictionHard.nodefs.inodesFree` parameter means that the Pods are evicted when more than 15% of the node's main filesystem's inodes are in use.
```

.Verification

> . You can check that your settings are in the `config.yaml` file in the `/var/lib/microshift/resources/kubelet/config/` directory.

Reviewer: Side note: this is after MicroShift starts. Not sure if we need to include it or it is obvious

Author: It is, but you never know.

> . After you complete the next steps and restart the host, you can use a root-access account to check that your settings are in the `config.yaml` file in the `/var/lib/microshift/resources/kubelet/config/` directory.

Reviewer 2: In the core OCP docs, one-step procedures take an asterisk. 

```suggestion
* After you complete the next steps and restart the host, you can use a root-access account to check that your settings are in the `config.yaml` file in the `/var/lib/microshift/resources/kubelet/config/` directory.
```

.Next steps
. Enable workload partitioning.
. Tune your operating system. For example, configure and activate a TuneD profile.
. Optional: Configure automatic enablement of your TuneD profile.
. Optional: If you are using the x86_64 architecture, you can install {op-system-rt-kernel}.

> . Prepare your {microshift-short} workloads for low latency.

Reviewer 2: Do we want to include links to make sure folks go to the right places?

Author: Would if I could!
